In [2]:
# Core libraries for data manipulation
import pandas as pd
import numpy as np

# Display settings so we can see all columns in outputs
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 200)

# Suppress unnecessary warnings for now
import warnings
warnings.filterwarnings('ignore')

print("Libraries loaded ✅")

Libraries loaded ✅


In [3]:
# Load CSV from the raw data folder
df = pd.read_csv('../data/raw/city_day.csv')

# Confirm it loaded
print("Dataset loaded successfully ✅")
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")

Dataset loaded successfully ✅
Rows: 29531, Columns: 16


In [4]:
# Look at the first 5 rows to see actual values
df.head()

,City,Date,PM2.5,PM10,NO,NO2,NOx,NH3,CO,SO2,O3,Benzene,Toluene,Xylene,AQI,AQI_Bucket
0,Ahmedabad,2015-01-01,NaN,NaN,0.92,18.22,17.15,NaN,0.92,27.64,133.36,0.00,0.02,0.00,NaN,NaN
1,Ahmedabad,2015-01-02,NaN,NaN,0.97,15.69,16.46,NaN,0.97,24.55,34.06,3.68,5.50,3.77,NaN,NaN
2,Ahmedabad,2015-01-03,NaN,NaN,17.40,19.30,29.70,NaN,17.40,29.07,30.70,6.80,16.40,2.25,NaN,NaN
3,Ahmedabad,2015-01-04,NaN,NaN,1.70,18.48,17.97,NaN,1.70,18.59,36.08,4.43,10.14,1.00,NaN,NaN
4,Ahmedabad,2015-01-05,NaN,NaN,22.10,21.42,37.76,NaN,22.10,39.33,39.31,7.01,18.89,2.78,NaN,NaN


In [5]:
# 1. Shape
print(f"Shape: {df.shape}")

# 2. All column names
print("\nColumns:")
print(df.columns.tolist())

# 3. Data types + non-null counts (this is gold)
print("\nData Info:")
df.info()

Shape: (29531, 16)

Columns:
['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket']

Data Info:
<class 'pandas.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  str    
 1   Date        29531 non-null  str    
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null

In [6]:
# Describe gives min, max, mean, quartiles for numeric columns
df.describe().T   # .T transposes for easy reading

,count,mean,std,min,25%,50%,75%,max
PM2.5,24933.0,67.450578,64.661449,0.04,28.820,48.57,80.5900,949.99
PM10,18391.0,118.127103,90.605110,0.01,56.255,95.68,149.7450,1000.00
NO,25949.0,17.574730,22.785846,0.02,5.630,9.89,19.9500,390.68
NO2,25946.0,28.560659,24.474746,0.01,11.750,21.69,37.6200,362.21
NOx,25346.0,32.309123,31.646011,0.00,12.820,23.52,40.1275,467.63
NH3,19203.0,23.483476,25.684275,0.01,8.580,15.85,30.0200,352.89
CO,27472.0,2.248598,6.962884,0.00,0.510,0.89,1.4500,175.81
SO2,25677.0,14.531977,18.133775,0.01,5.670,9.16,15.2200,193.86
O3,25509.0,34.491430,21.694928,0.01,18.860,30.84,45.5700,257.73
Benzene,23908.0,3.280840,15.811136,0.00,0.120,1.07,3.0800,455.03


In [7]:
# Missing value count + percentage, sorted high to low
missing = df.isnull().sum().sort_values(ascending=False)
missing_pct = (missing / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Missing Count': missing,
    'Missing %': missing_pct
})

# Show only columns that HAVE missing values
missing_df[missing_df['Missing Count'] > 0]

,Missing Count,Missing %
Xylene,18109,61.32
PM10,11140,37.72
NH3,10328,34.97
Toluene,8041,27.23
Benzene,5623,19.04
AQI,4681,15.85
AQI_Bucket,4681,15.85
PM2.5,4598,15.57
NOx,4185,14.17
O3,4022,13.62


In [8]:
# How many unique cities?
print(f"Unique cities: {df['City'].nunique()}")
print("\nCity list:")
print(sorted(df['City'].unique()))

# Date range (temporarily convert for inspection)
df['Date'] = pd.to_datetime(df['Date'])
print(f"\nDate range: {df['Date'].min()} → {df['Date'].max()}")
print(f"Total days covered: {(df['Date'].max() - df['Date'].min()).days} days")

Unique cities: 26

City list:
['Ahmedabad', 'Aizawl', 'Amaravati', 'Amritsar', 'Bengaluru', 'Bhopal', 'Brajrajnagar', 'Chandigarh', 'Chennai', 'Coimbatore', 'Delhi', 'Ernakulam', 'Gurugram', 'Guwahati', 'Hyderabad', 'Jaipur', 'Jorapokhar', 'Kochi', 'Kolkata', 'Lucknow', 'Mumbai', 'Patna', 'Shillong', 'Talcher', 'Thiruvananthapuram', 'Visakhapatnam']

Date range: 2015-01-01 00:00:00 → 2020-07-01 00:00:00
Total days covered: 2008 days


In [9]:
# See what AQI_Bucket categories exist
print("AQI_Bucket value counts:")
print(df['AQI_Bucket'].value_counts(dropna=False))

AQI_Bucket value counts:
AQI_Bucket
Moderate        8829
Satisfactory    8224
NaN             4681
Poor            2781
Very Poor       2337
Good            1341
Severe          1338
Name: count, dtype: int64


In [10]:
# 1. Any fully-duplicated rows?
print(f"Duplicate rows: {df.duplicated().sum()}")

# 2. Are there (City, Date) duplicates? (should be unique)
print(f"Duplicate (City, Date) pairs: {df.duplicated(subset=['City','Date']).sum()}")

# 3. Sanity check AQI: any negative values?
print(f"Negative AQI rows: {(df['AQI'] < 0).sum()}")

# 4. What's the max AQI row look like?
print("\nRow with max AQI:")
print(df.loc[df['AQI'].idxmax()][['City', 'Date', 'PM2.5', 'PM10', 'AQI', 'AQI_Bucket']])

Duplicate rows: 0
Duplicate (City, Date) pairs: 0
Negative AQI rows: 0

Row with max AQI:
City                    Ahmedabad
Date          2018-02-19 00:00:00
PM2.5                      242.66
PM10                          NaN
AQI                        2049.0
AQI_Bucket                 Severe
Name: 1145, dtype: object


## 📊 Data Understanding Summary

- **Shape:** 29,531 rows × 16 columns
- **Cities:** 26
- **Date range:** 2015-01-01 → 2020-07-01 (5.5 years)
- **Target:** AQI (numeric) / AQI_Bucket (category)
- **Key pollutant features:** PM2.5, PM10, NO2, CO, SO2, O3
- **Heavily missing (>80%):** NO, NO2, NOx → drop from features
- **Moderately missing (15–40%):** PM10, SO2, PM2.5, AQI → impute (pollutants) / drop (AQI)
- **AQI missing:** 4,682 rows (15.85%) → will drop
- **Extreme AQI:** 2049 on 2016-11-06 in Delhi (real event — keep, but flag)
- **No duplicate rows.**